pandas/tests/io/parser/test_network.py

```diff
@@ -188,3 +182,22 @@ def test_read_csv_handles_boto_s3_object(self,
 
         expected = read_csv(tips_file)
         tm.assert_frame_equal(result, expected)
+
+    def test_read_csv_chunked_download(self, s3_resource, caplog):
+        # 8 MB, S3FS usees ~5MB chunks
+        df = DataFrame(np.random.randn(100000, 4), columns=list('abcd'))
+        buf = BytesIO()
+        str_buf = StringIO()
+
+        df.to_csv(str_buf)
+
+        buf = BytesIO(str_buf.getvalue().encode('utf-8'))
+
+        s3_resource.Bucket("pandas-test").put_object(
+            Key="large-file.csv",
+            Body=buf)
+
+        with caplog.at_level(logging.DEBUG, logger='s3fs.core'):
```

Comment:
@martindurant does this seem like a reasonable way to test that `read_csv(key, nrows=5)` only triggers S3FS reading part of the object? Do you know of a better way, that's perhaps less reliant on the internals of S3FS?

Reply:
I'm afraid I don't have a better method for you, s3fs doesn't keep a log of transactions in any data structure you could access, and the s3file used for the download will have been cleaned up as soon as read_csv is done with it.

Reply:
Hmm OK. Do you know if moto keeps a record anywhere?

I dislike this test since s3fs adding an additional `logger.debug` anywhere, or changing the log message, default bytes size, etc. will break it.

boto also has a callback mechanism on `download_file`, but I don't see that option for `get_object`. If I can't figure out a way to get that working, I'll try to make the test using the logger a bit less fragile.

Reply:
Certainly you could look through all log messages captured, not just the last one. Note that you *do* have access to the exact s3filesystem in S3FileSystem._singleton[0], but I don't see that that helps you in this case.

You could maybe patch `S3File.__exit__` to store the values of `self.loc` and `self.end`?